src/setindex.jl

```diff
@@ -5,7 +5,9 @@ end
 @inline setindex(::Base.RefValue, val) = Ref(val)
 
 Base.@propagate_inbounds function setindex(xs::AbstractArray, v, I...)
-    T = promote_type(eltype(xs), typeof(v))
+    T = all(i -> i isa Integer, I) ?
```

Comment:
Not sure if you are aware, @phipsgabler works on the same issue in Setfield.jl https://github.com/jw3126/Setfield.jl/issues/175 . Your approach looks very slick, but it might miss some corner cases. For instance 
```julia
@set v[CartesianIndex(1)] = :hello
```
will produce a `Vector{Any}` I think?

Reply:
Funny, I came up with the exact same solution yesterday, seems like it is a sensible one :D But is your way type stable? I out of habit put the check and selection into a generated function.

Reply:
I have proposed two implementations in the issue linked above. But the `I isa Tuple{Vararg{Integer}}` variant seems superior to both.